core/arch/atomic_exports.h

```diff
@@ -723,7 +723,296 @@ atomic_dec_becomes_zero(volatile int *var)
+#        define ATOMIC_DEC(type, var) ATOMIC_DEC_##type(&var)
+
+#        undef ADDI_suffix
+
+static inline void
+ATOMIC_ADD_int(volatile int *var, int val)
+{
+    /* Page 53 of Volume I: RISC-V Unprivileged ISA V20191213 requires that var address is
+     * naturally aligned.
+     */
+    ASSERT(ALIGNED(var, 4));
+    __asm__ __volatile__("amoadd.w.aqrl zero, %1, 0(%0)\n\t"
+                         :
+                         : "r"(var), "r"(val)
+                         : "memory");
+}
+
+static inline void
+ATOMIC_ADD_int64(volatile int64 *var, int64 val)
+{
+    /* Page 53 of Volume I: RISC-V Unprivileged ISA V20191213 requires that var address is
+     * naturally aligned.
+     */
+    ASSERT(ALIGNED(var, 8));
+    __asm__ __volatile__("amoadd.d.aqrl zero, %1, 0(%0)\n\t"
+                         :
+                         : "r"(var), "r"(val)
+                         : "memory");
+}
+
+#        define ATOMIC_ADD(type, var, val) ATOMIC_ADD_##type(&var, val)
+
+static inline int
+atomic_add_exchange_int(volatile int *var, int val)
+{
+    int res;
+    /* Page 53 of Volume I: RISC-V Unprivileged ISA V20191213 requires that var address is
+     * naturally aligned.
+     */
+    ASSERT(ALIGNED(var, 4));
+    __asm__ __volatile__("amoadd.w.aqrl %0, %2, 0(%1)\n\t"
+                         : "=r"(res)
+                         : "r"(var), "r"(val)
+                         : "memory");
+    return res + val;
+}
+
+static inline int64
+atomic_add_exchange_int64(volatile int64 *var, int64 val)
+{
+    int64 res;
+    /* Page 53 of Volume I: RISC-V Unprivileged ISA V20191213 requires that var address is
+     * naturally aligned.
+     */
+    ASSERT(ALIGNED(var, 8));
+    __asm__ __volatile__("amoadd.d.aqrl %0, %2, 0(%1)\n\t"
+                         : "=r"(res)
+                         : "r"(var), "r"(val)
+                         : "memory");
+    return res + val;
+}
+
+#        define atomic_add_exchange atomic_add_exchange_int
+
+static inline bool
+atomic_compare_exchange_int(volatile int *var, int compare, int exchange)
+{
+    bool res = 1;
+    /* Page 49 of Volume I: RISC-V Unprivileged ISA V20191213 requires that var address is
+     * naturally aligned.
+     */
+    ASSERT(ALIGNED(var, 4));
+    /* This code is based on GCC's atomic_compare_exchange_strong().
+     * The fence ensures no re-ordering beyond the start of the lr/sc loop.
+     * The lr/sc use acquire ordering to ensure that the full loop has
+     * sequentially consistent memory order semantics.
+     */
+    __asm__ __volatile__("fence   iorw,ow\n\t"
+                         "1: lr.w.aq %0, (%1)\n\t"
+                         "   bne     %0, %2, 2f\n\t"
+                         "   sc.w.aq %0, %3, (%1)\n\t"
+                         "   bnez    %0, 1b\n\t"
+                         "2:"
+                         : "+r"(res)
+                         : "r"(var), "r"(compare), "r"(exchange)
+                         : "memory");
+    return res == 0;
```

Comment:
atomic_compare_exchange_int is supposed to return true if var was equal to compare.

If they were equal, res will have 0 indicating success of the store conditional, so we'll return true, which is correct.

If they weren't equal we'd branch to `2:` and res will have whatever value was in val. If it was zero, we'll incorrectly return true.